packages/model-viewer/src/three-components/gltf-instance/VariantMaterialLoaderPlugin.ts

```diff
@@ -114,38 +114,28 @@ export default class GLTFMaterialsVariantsExtension implements
     for (const scene of gltf.scenes) {
       // Save the variants data under associated mesh.userData
       scene.traverse(object => {
-        // The following code can be simplified if parser.associations directly
-        // supports meshes.
-        const association = parser.associations.get(object);
+        const mesh = object as Mesh;
 
-        if (association == null || association.meshes == null) {
+        if (!mesh.isMesh) {
           return;
         }
 
-        const meshIndex = association.meshes;
+        const association = parser.associations.get(mesh) as GLTFReference & {primitives: number};
 
-        // Two limitations:
-        // 1. The nodeDef shouldn't have any objects (camera, light, or
-        // nodeDef.extensions object)
-        //    other than nodeDef.mesh
-        // 2. Other plugins shouldn't change any scene graph hierarchy
-        // The following code can cause error if hitting the either or both
-        // limitations If parser.associations will directly supports meshes
-        // these limitations can be removed
+        if (association == null || association.meshes == null || association.primitives == null) {
+          return;
+        }
 
-        const meshDef = json.meshes[meshIndex];
+        const meshDef = json.meshes[association.meshes];
         const primitivesDef = meshDef.primitives;
-        const meshes = 'isMesh' in object ? [object] : object.children;
-
-        for (let i = 0; i < primitivesDef.length; i++) {
-          const primitiveDef = primitivesDef[i];
-          const extensionsDef = primitiveDef.extensions;
-          if (!extensionsDef || !extensionsDef[this.name]) {
-            continue;
-          }
-          meshes[i].userData.variantMaterials =
-              mappingsArrayToTable(extensionsDef[this.name]);
+        const primitiveDef = primitivesDef[association.primitives];
+        const extensionsDef = primitiveDef.extensions;
+
+        if (!extensionsDef || !extensionsDef[this.name]) {
+          return;
         }
+
+        mesh.userData.variantMaterials = mappingsArrayToTable(extensionsDef[this.name]);

```

Comment:
This is much simpler, which I like! I think this works now because of upstream changes in three.js; I believe `association.primitives` didn't exist at the time this was written. Thanks!